node-it/src/test/scala/com/wavesplatform/it/sync/AddressApiSuite.scala

```diff
@@ -33,10 +46,89 @@ class AddressApiSuite extends BaseTransactionSuite {
         sender.getData(firstAddress, invalidRegexp)
         fail("RegexCompiler didn't throw expected error")
       } catch {
-        case err: Throwable => if (!err.getMessage.contains("Cannot compile regex")) {
-          throw err
-        }
+        case err: Throwable =>
+          if (!err.getMessage.contains("Cannot compile regex")) {
+            throw err
+          }
       }
     }
   }
+
+  test("balances for waves should be correct") {
+    assertBalances(None)
+  }
+
+  test("balances for issued asset should be correct") {
+    val asset = miner.issue(miner.address, "Test", "Test", 10000, 0, waitForTx = true).id
+    assertBalances(Some(asset))
+  }
+
+  test("limit violation requests should be handled") {
+    val limit     = miner.config.getInt("waves.rest-api.transactions-by-address-limit")
+    val address   = miner.createAddress()
+    val addresses = List.fill(limit + 1)(address)
+    assertApiError(
+      miner.get(s"/addresses/balance?${addresses.map(a => s"address=$a").mkString("&")}"),
+      TooBigArrayAllocation
+    )
+    assertApiError(
+      miner.postJson(
+        "/addresses/balance",
+        Json.obj("ids" -> addresses)
+      ),
+      TooBigArrayAllocation
+    )
+  }
+
+  private def assertBalances(asset: Option[String]): Unit = {
+    val addresses = (1 to 5).map(i => (miner.createAddress(), (i * 100).toLong)).toList
+
+    val firstAddresses  = addresses.slice(0, 2)
+    val secondAddresses = addresses.slice(2, 5)
+
+    val heightBefore  = transferAndReturnHeights(firstAddresses, asset).min - 1
+    val heightBetween = nodes.waitForHeightArise()
```

Comment:
use sender.waitForHeight(heightBefore+3)

Reply:
@DrBlast Theoretically it can be that sent transactions mined in different blocks, so it is not correct to rely on the `heighBefore`